holoviews/plotting/bokeh/element.py

```diff
@@ -887,25 +887,34 @@ def _get_colormapper(self, dim, element, ranges, style):
                 return cmapper
             else:
                 return None
-        colors = self.clipping_colors
-        if isinstance(low, (bool, np.bool_)): low = int(low)
-        if isinstance(high, (bool, np.bool_)): high = int(high)
-        opts = {'low': low, 'high': high}
-        color_opts = [('NaN', 'nan_color'), ('max', 'high_color'), ('min', 'low_color')]
-        for name, opt in color_opts:
-            color = colors.get(name)
-            if not color:
-                continue
-            elif isinstance(color, tuple):
-                color = [int(c*255) if i<3 else c for i, c in enumerate(color)]
-            opts[opt] = color
+
+        ncolors = None if factors is None else len(factors)
+        low, high = ranges.get(dim.name, element.range(dim.name))
+        palette = mplcmap_to_palette(style.pop('cmap', 'viridis'), ncolors)
+
+        color_rgb = lambda color: [int(c*255) if i<3 else c for i, c in enumerate(color)]
+        colors = {k: color_rgb(v) if isinstance(v, tuple) else v
+                  for k, v in self.clipping_colors.items()}
+
+        if factors is None:
+            colormapper = LogColorMapper if self.logz else LinearColorMapper
+            if isinstance(low, (bool, np.bool_)): low = int(low)
+            if isinstance(high, (bool, np.bool_)): high = int(high)
+            opts = {'low': low, 'high': high}
+            color_opts = [('NaN', 'nan_color'), ('max', 'high_color'), ('min', 'low_color')]
+            opts.update({opt: colors[name] for name, opt in color_opts if name in colors})
+        else:
+            colormapper = CategoricalColorMapper
+            opts = dict(factors=factors)
+            if 'NaN' in colors:
+                opts['nan_color'] = colors['NaN']
+
```

Comment:
I am wondering if a chunk of this should be a utility, e.g [these lines](https://github.com/ioam/holoviews/blob/77f14160c249b0fcced16ffe31f7d2e00c942e7b/holoviews/plotting/bokeh/element.py#L891-L910). Both ``self.clipping_colors`` and ``self.logz`` seem to be sensible argument for a general color mapping utility.

I'll leave it up to you but to me it seems like a fair chunk of code that doesn't have to be tied up in a plotting class.